### DISCUSSION TITLE
Go lang vs Rust for Different Use Cases

### DISCUSSION THREAD

**Discussion by alimd** [AliMD/alwatr#126] [Educational]
We need to know when to invest in `Go` and when to invest in `Rust` and why?
Please comment everything you found. Thank you.

**Comment by alimd** [comment]
از مقالات ساده و ابتدایی شروع کنیم

> Rust is great for building things like operating systems, file systems, and game engines. Go is best-suited for applications involving big data, machine learning, and editing massive files.

https://getclockwise.com/blog/rust-vs-go

**Comment by alimd** [comment]
بررسی مراجع و مخلفات اطراف هر زبان

[Awesome Rust](https://github.com/rust-unofficial/awesome-rust#readme)

[Awesome Go](https://github.com/avelino/awesome-go#readme)

**Comment by alimd** [comment]
مقاله ای مقدماتی ولی براش شروع مفید

https://bitfieldconsulting.com/golang/rust-vs-go

**Comment by alimd** [comment]
یک نمونه از کاربردهایی که راست خیلی قدرت نمایی می‌کنه و در دنیای سرویس‌های کلود وسوسه کننده است

https://github.com/joseluisq/static-web-server

حالا سوال اینجاست آیا از ابزار های موجود مثل Nginx خیلی سریعتره؟

**Comment by alimd** [comment]
پروژه‌ای که بهت احساس قدرت و بی سوادی شدید میده
https://tokio.rs/

**Comment by alimd** [comment]
در عظمت Rust شکی نیست ولی یک نکته جالب!

با تجربه ای که در شیوه کد نویسی بزرگان داریم از نکات Clean Code تا نظم در فایل ها تا مباحث ساده ای مثل کامنت نویسی اصولی، سورس این دوتا پروژه رو باهم مقایسه کنید!

https://github.com/golang/go/tree/master/src
vs
https://github.com/rust-lang/rust/tree/master/src

سورس Go به شدت اصولی و تمیز هستش
کاملا ساختار مند، همه چی خیلی تمیز و اصولی کامنت داره، فایل ها خیلی منظم دسته بندی و مفهومی فولدر بندی شده و خلاصه کاملا مشخص هستش که یک کمپانی بزرگ و Enterprise نوشته‌اش

ولی سورس Rust به ظاهر خیلی شلخته و نا مرتب و بدون حتی کامنت هستش
انقدر کثیف و نامرتب هستش که مجدد چک کردم ببینم این سورس واقعا همون زیان بزرگ هستش!
کاملا احساس میکنی توسط یک آدم معمولی و نه چندان باسواد نوشته شده (در صورتی که میدونیم اینطور نیست)
حتی گیت کامیت ها شلخته هستند، حتی پروژه‌های دیگه خیلی غیر حرفه‌ای git submodule شده! (هیچ پروژه خفنی نمیاد از submodule استفاده کنه)

دقیقا این نکات هستش که من رو نگران میکنه و باعث میشه بگم صبر کن یک بار دیگه همه چیز رو بررسی کن!

**Comment by njfamirm** [comment]


**Reply by alimd** [reply]
نه اصلا
دنیای Deno هیچ ربطی به کاربردهای Rust نداره
و هیچ موقع قرار نیست به سرعت Native برسه
این سوال رو برای این می‌پرسم که بتونم تصمیم درستی برای آینده م مخصوصا پروژه‌های کلودمون بگیرم
مثلا آیا ارزش داره میکروسرویس Data Storage برای 1DB رو با Rust بنویسم؟
من حس میکنم تو کلود ما به محدودیت سرعت رم، هارد و شبکه میخوریم، نیاز نداریم یک کار محاسباتی سنگین انجام بدیم که ارزش داشته باشه با Go یا Rust نوشته بشه
پس دقیقا کجا و چقدر باید تو این زبان‌ها سرمایه‌گذاری کرد

**Reply by njfamirm** [reply]
بنظرم تو وویس اینو گفتین 😆

**Reply by alimd** [reply]
چرا کامنت رو پاک کردی 
نکته مهمی بود بزار بمونه

**Comment by njfamirm** [comment]
بنظرم خوبه به زبان‌های استفاده شده در پروژه‌های فارغ التحصیل شده‌ی [cncf](https://www.cncf.io/projects/) نگاهی بیندازیم...

# Go

- containerd
- coredns
- etcd
- harbor
- helm
- kubernetes
- jaeger
- linkerd2
- opa
- prometheus
- rook

# rust
- tikv

# other
- python-tuf
- envoy
- fluented

این روند توی بقیه پروژه‌های cncf هم ادامه داره

**Reply by njfamirm** [reply]
با توجه به این موارد جامعه کلود کامیونیتی بزرگتری برای Go داره نسبت به Rust

**Reply by njfamirm** [reply]
همچنین API های این پروژه ها هم با Go کار می‌کنه مثلا `containerd`

**Reply by alimd** [reply]
در api ما پرفرمنس نمیخوایم برای همین خیلی جاها حتی پایتون هستش!

**Reply by njfamirm** [reply]
منظورم [چنین](https://github.com/containerd/containerd#features) چیزی بود
یجورایی منظورم راه های استفاده از یک پکیج
مگه به اینا هم نمیگفتن API?!

**Reply by alimd** [reply]
تو همچین مثال هایی اصلا پرفورمنس مهم نیست!
چون فقط جنبه کنترلی داره

**Reply by njfamirm** [reply]
خب  بله
حرفم این بود که آیا با دنو میشه به کانتینر دی وصل شد؟

**Comment by njfamirm** [comment]
توی ورژن 1.8 Go
جنریک هم اضافه شده
یعنی یکم از اون سخت‌گیری‌های قبلی برای تایپ‌ها کم شده

**Comment by njfamirm** [comment]
![image](https://user-images.githubusercontent.com/83071310/160274685-79109d35-ed6a-4a13-8c42-02470ad11ff8.png)

**Reply by alimd** [reply]
واقعا lamda expersion آورده!!!
وای خدای من
چرا احساس میکنم go 1 مال خودشون بود و go 2  مال community
مثل لیت و انگولار
خخخ

**Comment by njfamirm** [comment]
نتیجه گیری خودم اینکه
- با توجه به گفته‌ی تقریبا همه‌ی منابع Rust سریع‌تر از Go هست

- سینتکس Go هم یقینا از Rust ساده تره 
و با توجه به اضافه شدن جنریک به Go سخت‌گیری برای تایپ‌ها هم خیلی کمتر شده

- کامیونیتی Go هم بیشتره مخصوصا در cloud

همچنین فیچر های GO2 هم باید در نظر بگیریم

**Reply by alimd** [reply]
پس همون نتیجه قبلی
هزینه پایین توسعه در مقابل سرعت native 
پس اگر قراره هزینه توسعه ندی من با همون زبان هایی که خودم بهش مسلط هستم راحت ترم
مرسی

**Reply by njfamirm** [reply]
یعنی سوییچ میکنین به Go ?

**Reply by alimd** [reply]
اما این سوال موند
یک وب سرور با راست و گو چقدر تفاوت سرعت داره
از طرفی هیچ منبعی معرفی نکردی ها!

**Reply by njfamirm** [reply]
منابع همه یک چیزو میگن
سرعت راست
سینتکس ساده تر گو

**Reply by alimd** [reply]
> یعنی سوییچ میکنین به Go ?

نه دیگه من به go تسلطی ندارم
اگر قرار باشه هزینه توسعه و تسلط به زبان جدیدی رو بدیم طبیعتا به راست بدیم بهتره
ولی go خیلی در ذهنم جایگاه بهتری پیدا کرد

**Reply by asmanp2012** [reply]
> اما این سوال موند یک وب سرور با راست و گو چقدر تفاوت سرعت داره از طرفی هیچ منبعی معرفی نکردی ها!

خیلی در باره این موضوع سرچ کردم و خوشحالم که  نتایجش با عقایدی که داشتم یکی بودش از نظر من زبان عاملی برای انتخاب ما  نیستش و همه تکنولوژی ها میتونن قابلیت های فوقالعاده ای رو به ما بدن برای همین شروع کردم به بررسی برای این که آیا واقعا اگر از rust به عنوان  یک وب سرور استفاده کنیم آیا بهتر از nginx هستش ؟

اولین چیزی که بررسی کردم مقایسه nginx با nodjs بودش و تفاوت پرفورمنسشون منابع زیادی رو  پیدا نکردم بجز این که همه چیز رو روی نمودار آورده بود و تفاوت  اکسپرس با  nginx  بود

https://stackoverflow.com/a/16635547 - nginx vs express

خوب اگر نمودارش و بررسی کنیم میبینم با توجه به این که اکسپرس خودش تاخیر های زمانی و ... زیادی داره و اگر تفاوت  nodejs  با express رو در نظر بگیرم احساس میکنم میتونن برابر باشن و خیلی تفاوت آنچنانی ندارن و خوب از اونجا که منبع خیلی خوبی نیستش شاید شبهه ای توش باشه که خوب در ادامه بازم بررسیش میکنیم


بعد از اون رفتم سراغ بنچ مارک ها تا ببینم بهترین فریمورکی که از لحاظ پرفورمنسی وجود داره چی هستش و به این سایت رجوع کردم:

https://www.techempower.com/benchmarks/#section=data-r20&hw=ph&test=fortune

همینطور که میبینید دراگون کور بیشتری بازدهی رو نسبت به همه زبان ها داشته پس  دنبال تفاوت های nginx با drogon-core گشتم که خوشبختانه یه نفر این کار رو انجام داده بود:
https://drogon.docsforge.com/master/benchmarks/#test-plan-and-results

توی این بنچ مارک که توی سایت خوده دراگون هستش نشون میده که دراگون خیلی سریعتر از nginx هستش پس actix-core  هم میتونه پرفورمنسی نزدیک به dragon داشته باشه

اما وقتی حالت بدون هیچ دیتابیس و ... فریمورک رو توی همون بنچ مارک techempower چک میکنید میبیند که همه چیز تغیر میکنه زبان های مختلف همه دوشادوش هم دارن پرفورمنس یکسانی رو ارائه میدن  برای جاوا اسکریپت هم  (just-js) تو سطح  rust داره کارشو انجام میده گرچه تو گزینه های  دیگه هم همینطوریه ولی در حال حاضر صحبت ما سرو فایل  و ... هستش

https://www.techempower.com/benchmarks/#section=data-r20&hw=ph&test=plaintext

نتیجه من از این موضوع توی موضوع سرو فایل ها اینه که از اینجا به بعد فقط بسته به سیستم عامل  و سخت افزار هستش که چه میزان درخواستی قراره جواب داده بشه و اونجا باید دنبال جواب باشیم تو این لایه عمیق تر شدن صرفا میکرو اپتیمیزیشن های خیلی خیلی کوچیکی هستش که توی کل ماجرا تغیری نمیده و نگهداشتنش توی یک وضعیت خاص خیلی سخته و موضوع بعدی هم اینه که اگر قراره سرو فایل رو انجام بدیم بهتره که توی لایه هاش شبکه گمش نکنیم و پروکسی های مختلف رو سر راهش قرار ندیم و اگر کسی قراره این کار رو بکنه مستقیم این کار رو بکنه.

**Comment by njfamirm** [comment]
https://qvault.io/rust/rust-vs-go-in-backend-web-development/

**Reply by alimd** [reply]
Nice 👍 

## Rust is Favored For…
* Computation Speed
* Memory usage
* Fine-tuning and control over the final executable

## Go is Favored For…
* Simple syntax
* Beginner-friendly
* Developer speed
* Concurrency mechanisms

**Comment by njfamirm** [comment]
به نتیجه رسیدیم دیگه
